### DISCUSSION TITLE
Discussing multi-factor decryption and optional-factor decryption

### DISCUSSION THREAD

**Discussion by ashleysommer** [keepassxreboot/keepassxc#9606] [Ideas]
This discussion has evolved out of comments on the FIDO2 topic: https://github.com/keepassxreboot/keepassxc/discussions/9506, particularly this comment by @BryanJacobs https://github.com/keepassxreboot/keepassxc/discussions/9506#discussioncomment-6137689
Also probably closely related to the KDBX features topic: https://github.com/keepassxreboot/keepassxc/discussions/6229

For years there have been requests from users like "I want to use a password _or_ a keyfile to unlock the database" or "I want to use my password plus one of either my yubikey or my keyfile" or "I want to use my password and this yubikey, or my password and that solokey" or many others along those lines. Normally the answer is simply "KeepassXC doesn't support that, the KDBX format doesn't work that way, you cannot do that". 

In the topic of implementing FIDO2 key support, one very important aspect came up. Most other applications that use the FIDO2 protocol (like Webauthn, etc) allow you to register more than one FIDO2 device to use. The premise is it is easy to imagine losing a security key, or not having access to it when you need to unlock your database, so having a second device registered means you have a backup. We get around this issue with Yubikey support through the ability to clone the HMAC code from one yubkey to another, so it can provide the same challenge response. The FIDO2 hardware keys do not allow this, their memory is read-only and their HMAC code cannot be cloned. So the only way to have more than one FIDO2 hardware key to be used to unlock the database is to have optional factors supported. It was also expressed by some that it is not even worth implementing support for FIDO2 keys if it is not possible to enroll more than one key.

When discussing this in the [FIDO2 topic](https://github.com/keepassxreboot/keepassxc/discussions/9506), I had originally intended to implement this multiple-FIDO2-keys feature in a way that is contained within the FIDO2 implementation, with this ability to use one-of-many keys specific to the FIDO2 feature only. This is to minimise the impact this feature has on other functions of the database, and other decryption factors.

However, it was pointed out that this should be done in a more general way, implementing a new `AdditionalFactors` structure in the KDBX extended header that will allow support for all decryption factors (Password, Keyfile, Yubikey, and FIDO2), as well as enrol more than one of each, and have some optional sets (eg, 1 of these two Keyfiles, and 1 of these two FIDO2 keys). This would address the need to have this feature for optional FIDO2 keys, as well as open up the opportunity to address the older feature requests that I mentioned at the start. For more details, [see the comprehensive write up](https://github.com/keepassxreboot/keepassxc/discussions/9506#discussioncomment-6137689) by @BryanJacobs.

As [discussed in the other thread](https://github.com/keepassxreboot/keepassxc/discussions/9506#discussioncomment-6137833) this more generic method would not necessarily constitute more work than implementing it specific to FIDO2 keys as part of the FIDO2 feature implementation, but it does mean this change would be out of scope of that feature set, and will require much more discussion and direction. That is what this new discussion topic is opened for.

**Comment by droidmonkey** [comment]
> When the DB is unlocked, if the header contains 1-of-N, DO NOT ROTATE THE MASTER SECRET PART as you can't re-encrypt the other authenticators' versions of it. 

This is what is going to kill this effort. You will corrupt the use of the additional factors the moment you save it in any other program or a previous version of KeePassXC.

You are best to NOT do this and instead advocate for formal support of general multi-keying in KDBX 5.

**Reply by BryanJacobs** [reply]
I don't think an older version of keepass, or some other application, is likely to accidentally re-encrypt the stuff in the AdditionalFactors header, because it doesn't know how to write (or read) that structure.

Saving the database in an oblivious application won't "corrupt" the DB - it would just invalidate the 2FA. But to save the database you need to unlock it first... and how would that happen without 2FA?

The line above is meant to be a mandate for the new implementation of reading the 2FA headers, not something that needs to be heeded by older applications.

Because the way I designed the 2FA header it all just puts parts together to form the master secret, I don't see a risk of DB corruption. Could you explain in more detail what sequence of events you're worried about here?

**Reply by droidmonkey** [reply]
One of my edits didn't take. I meant to say another application saving that rotates the master secret would invalidate the additional credentials that relied on it being static.

**Reply by BryanJacobs** [reply]
That's correct, but luckily you need to use the 2FA credentials to open the database at all, right? And so you couldn't save it without implementing the spec?

I don't think being able to open and re-save the database without using the 2FA header is a useful thing to enable.

**Reply by droidmonkey** [reply]
Yeah, so that's the part I didn't get from your post. You are actually proposing a completely breaking change to the kdbx standard by making it impossible to open the database in an unaware application. We are accepting that situation with FIDO2 addition since it is deliberate to even support the keys in the first place. I would definitely not merge that behavior in a general sense. That is why this needs to be a deliberate change in kdbx 5 that all apps agree to.

Also, the spillage of meta data is a big concern of mine. Leaving a key file name in the open (at least hash it) or relying on a file name at all to correlate material is a non-starter. Also, tying the Yubikey hmac CR to serial number is the same problem. Right now kdbx is basically zero knowledge encryption. We have no idea what constitutes your key material until it actually works.

**Reply by BryanJacobs** [reply]
I'm definitely proposing a change that's not backwards-compatible, but it's no more (or less) of a "breaking change" than just adding FIDO2 keys would be - if you don't support the FIDO2 authenticators, you can't open the database. The same was done in KDBX v4 when yubikey-hmac support was added before. A database saved using a yubikey-hmac can't be opened by an application that doesn't support yubikey-hmac.

About putting authenticator meta-info into the header: I understand that point of view. Be aware that if you want it to be "zero knowledge" you will have to use FIDO2 resident credentials *only*. In order to support non-resident credentials somebody has to store a blob that the authenticator returns on registration.

Personally, I think the user convenience of knowing which authenticators to use outweighs the theoretical security benefit of having the database not say that. You can not-store file paths, but it'd be nice to know that the database requires a key file... Instead of, you know, just displaying an error that says the password was wrong.

Kerckhoff's Principle says that knowing HOW the database is encrypted shouldn't compromise its security at all. And if you're for reducing the risk of corruption, you're going to want to track whether multiple credentials are registered or not...

**Reply by droidmonkey** [reply]
Agreed, I am not against knowing the 2FA options in the metadata. That simplifies a lot of implementation and improves user experience without compromising security. Just be careful of strongly binding 2FA options to ephemeral/inconsequential attributes like file name or serial number. The exception of course is with FIDO2 which must correlate with the key serial number. 

If you can somehow pulloff the major ui changes necessary to make this pleasant while also very strongly warning users they will break compatibility with non-supportive apps.... I would consider it. However I would definitely prefer a kdbx version bump to make it obvious to other apps that you "cannot open this file".

**Reply by ashleysommer** [reply]
> When the DB is unlocked, if the header contains 1-of-N, DO NOT ROTATE THE MASTER SECRET PART as you can't re-encrypt the other authenticators' versions of it.

That's right, we cannot rotate the master secret part that is shared and separately encrypted by secrets from each authenticator, but we can rotate the salt (FIDO2's name for the challenge string) that is saved in the db and reencrypt the master key part for the currently used authenticator only. 

Hmm, that brings up another question. What happens if the DB needs to save, but the authenticator used to decrypt it has been removed from the device? Does KeepassXC need to deal with that in the current Yubikey implementation? I haven't checked what happens.

**Reply by droidmonkey** [reply]
We wait until the yubikey with the same serial that unlocked the database is available in order to save. This is mainly because the master seed (the thing that gets rotated on save) serves as the challenge. 

I would actually not use the master seed as a challenge in this 2FA implementation. Each factor should have its own seed that can be independently rotated. The master seed is used in the combination of key material, but not used when unlocking the "crypto box".

**Reply by ashleysommer** [reply]
> Each factor should have its own seed that can be independently rotated.

Yeah, that's what I was referring to above, in FIDO2 terms that is called the "HMAC Salt". That will be per-key and can be rotated each save. That means we would still need the key present at save time, just like with the Yubikey code.

**Reply by BryanJacobs** [reply]
If you use the HMAC response as two separate 128-bit salts (instead of one 256-bit), you don't have to present the key again at save time.

When unlocking the database send the key the current value, and also the value-you-will-use-for-saving. Then you can rotate the salt one time without needing the user to present the authenticator again.

I think it's okay that if the user edits the database and saves it five times in succession the salt is only rotated once, right?

**Reply by droidmonkey** [reply]
Yes, one rotation per 'session' makes a lot of sense.
